vault/data_source_namespaces.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/vault/api"
+
+	"github.com/hashicorp/terraform-provider-vault/internal/consts"
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+	"github.com/hashicorp/terraform-provider-vault/util/mountutil"
+)
+
+func namespacesDataSource() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: provider.ReadContextWrapper(namespacesDataSourceRead),
+
+		Schema: map[string]*schema.Schema{
+			consts.FieldPaths: {
+				Type:        schema.TypeSet,
+				Computed:    true,
+				Description: "Namespace paths.",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func namespacesDataSourceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	log.Printf("[DEBUG] Reading namespaces from Vault")
+
+	resp, err := client.Logical().ListWithContext(ctx, consts.SysNamespaceRoot)
+	if err != nil {
+		return diag.Errorf("error reading namespaces from Vault: %s", err)
+	}
+	if err := d.Set(consts.FieldPaths, flattenPaths(resp)); err != nil {
+		return diag.Errorf("error setting %q to state: %s", consts.FieldPaths, err)
+	}
+
+	id := mountutil.NormalizeMountPath(client.Namespace())
```

Comment:
Wondering if we should use a UUID here instead. Would this value be unique between two instances of this data source? IIUC this would set all instances of the data sources to the same ID, the namespace of the provider's client?

Reply:
Good question! Does a unique ID actually matter for a data source? In my tests I can't see any conflict or reason to change this. What happens if the ID is not unique?

In practical terms, I don't see a reason to have the exact same copy of this data source. For a given namespace level, the data will all be the same.

I don't feel too strongly about it either way. Are there any downsides to using a UUID instead?

Reply:
I ran a bit of test TF to look at the state, and you are correct. I realize that this shouldn't be a concern since we should only ever have one of these datasources for a particular `namespace` param. Another thing I was forgetting is that `provider.GetClient` appropriately fetches the correct namespaced client based on the `namespace` parameter provided. So for each instance of this data source in a `namespace`, the value of `client.Namespace()` should result in a unique value. I believe in general we always want an instance of resource/data source to have a unique ID, so I just wanted to give this a second thought. Thanks for indulging and apologies for any confusions! 🙏🏼

Reply:
No worries! I think this will conflict with the new `vault_namespace` (singular) data source. So maybe we should change it so it doesn't conflict with that?

Reply:
I think across resources and data sources ID uniqueness is not mandatorily enforced. In my understanding, it is only enforced across multiple instances of the same resource/datasource, so we should be good